### PR TITLE
Feature/246/password reset

### PR DESCRIPTION
**Pull request by andreaMT15**
### Checklist For Submitter

* [ ] I have documented all new functionality
* [ ] I have screenshotted new UI changes and attached them to this PR

### CODE REVIEW

File: config/environments/production.rb [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
 
+  config.action_mailer.smtp_settings = {
+    address: "smtp.gmail.com",
+    port: 587,
+    domain: "example.com",
+    authentication: "plain",
+    enable_starttls_auto: true,
+    user_name: ENV["GMAIL_USERNAME"],
+    password:ENV["GMAIL_PASSWORD"]
+  }
```
Comment (vhoof) [inline config/environments/production.rb]: I think we'll want to use Sendgrid for this kind of stuff, especially in production. In particular, we'll want to use `SENDGRID_USERNAME` and `SENDGRIND_PASSWORD`. We'll also want to `smtp.sendgrid.net` as the address, and something like `rescuingleftovercuisine.org` as the domain.

I think we should do this in staging, too, and we can sort out how to test it with .env.

Also, probably want a space after `password:`.

File: config/environments/production.rb [hunk 2]
```diff
@@ -1,6 +1,17 @@
...
+
+  config.action_mailer.default_url_options = { :host => 'yourhostsite.herokuapp.com' }
```
Comment (vhoof) [inline config/environments/production.rb]: We'll probably want this to be `ENV['CLIENT_URL']`.

File: config/environments/development.rb [hunk 3]
```diff
@@ -4,6 +4,18 @@
...
+  config.action_mailer.smtp_settings = {
+    address: "smtp.gmail.com",
+    port: 587,
+    domain: "example.com",
+    authentication: "plain",
+    enable_starttls_auto: true,
+    user_name: ENV["GMAIL_USERNAME"],
+    password:ENV["GMAIL_PASSWORD"]
+  }
+config.action_mailer.default_url_options = { :host => 'localhost:3000' }
```
Comment (vhoof) [inline config/environments/development.rb]: We'll want to tab this one over once, and we want to use the front-end, so port 5000.

File: frontend/src/components/NewPassword/NewPassword.js [hunk 4]
```diff
@@ -22,14 +27,43 @@ class NewPassword extends Component {
...
+          changeCreatedState(response.status);
+          history.push(queryString);
+        })
+        .catch((error) => {
+          if (error.message.includes('422')) {
+            const parentForm = document.getElementsByTagName('form')[0];
+            const errorDiv = document.createElement('p');
+            errorDiv.setAttribute('class', 'errorMessage reset-password');
+            errorDiv.innerHTML = 'Please enter a valid password.';
+            parentForm.insertAdjacentElement('beforeend', errorDiv);
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: If we're going to copy this from that other page, we should probably use the updated version from SignIn to avoid repeated errors.

Comment (vhoof) [reply frontend/src/components/NewPassword/NewPassword.js]: See https://github.com/the-difference-engine/rescue-leftover-cuisine2/blob/staging/frontend/src/components/SignIn/SignIn.js#L37-L40 and https://github.com/the-difference-engine/rescue-leftover-cuisine2/blob/staging/frontend/src/components/SignIn/SignIn.js#L97

File: frontend/src/containers/ResetPassword/ResetPasswordPage.js [hunk 5]
```diff
@@ -1,23 +1,30 @@
...
-    </div>
-    <div className="passwordBackgroundImage">
-      <div className="newPasswordCardWrapper">
-        <NewPassword />
-        <div className="loginFooter row">
-          <Footer />
+const ResetPasswordPage = (props) => {
+  const { location } = props;
+  const parsed = querystring.parse(location.search.slice(1));
+  const token = parsed.q;
```
Comment (vhoof) [inline frontend/src/containers/ResetPassword/ResetPasswordPage.js]: I'd probably use `token` as the query param, though we'll need to see what rails wants to do.

File: frontend/src/components/NewPassword/NewPassword.js [hunk 6]
```diff
@@ -48,13 +82,12 @@ class NewPassword extends Component {
...
           <div className="passwordConfirm form-group row">
             <input
               type={isPasswordVisible ? 'text' : 'password'}
               id="inputSignInPasswordConfirm"
               className="password-confirm-input fullWidth form-control-lg"
-              name="confirm-password"
+              name="confirmPassword"
               required
               placeholder="Confirm Password"
               onFocus={event => event.target.setAttribute('placeholder', 'confirmPassword')}
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: Can we change this placeholder to say 'Confirm Password'?

File: frontend/src/components/NewPassword/NewPassword.js [hunk 7]
```diff
@@ -22,14 +27,43 @@ class NewPassword extends Component {
...
+            errorDiv.innerHTML = 'Please enter a valid password.';
+            parentForm.insertAdjacentElement('beforeend', errorDiv);
+          }
+        });
+    }
+  };
+
   render() {
-    const { changeAuthorizedState } = this.props;
+    const { changeCreatedState } = this.props;
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: Doesn't look like this is being passed in, where does this come from/do?

File: frontend/src/components/NewPassword/NewPassword.js [hunk 8]
```diff
@@ -22,14 +27,43 @@ class NewPassword extends Component {
...
+    const { password, confirmPassword } = this.state;
+    const { resetPasswordToken } = this.state;
+    const queryString = `resetpassword/?q=${resetPasswordToken}`;
+    if (password !== confirmPassword) {
+      alert("Passwords don't match"); // eslint-disable-line no-undef
+    } else {
+      resetPassword(this.state)
+        .then((response) => {
+          changeCreatedState(response.status);
+          history.push(queryString);
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: I don't think we want to be redirecting from here to the reset password page, as that's where we are. We could just redirect to the homepage?

File: frontend/src/components/NewPassword/NewPassword.js [hunk 9]
```diff
@@ -22,14 +27,43 @@ class NewPassword extends Component {
     this.setState({ [event.target.name]: event.target.value });
   };
 
+  handleSubmit = (event, changeCreatedState) => {
+    event.preventDefault();
+    const { history } = this.props;
+    const { password, confirmPassword } = this.state;
+    const { resetPasswordToken } = this.state;
+    const queryString = `resetpassword/?q=${resetPasswordToken}`;
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: I don't think we need this.

File: frontend/src/containers/ResetPassword/ResetPasswordPage.js [hunk 10]
```diff
@@ -1,23 +1,30 @@
...
-    </div>
-    <div className="passwordBackgroundImage">
-      <div className="newPasswordCardWrapper">
-        <NewPassword />
-        <div className="loginFooter row">
-          <Footer />
+const ResetPasswordPage = (props) => {
+  const { location } = props;
+  const parsed = querystring.parse(location.search.slice(1));
+  const token = parsed.t;
```
Comment (andreaMT15) [inline frontend/src/containers/ResetPassword/ResetPasswordPage.js]: @vhoof in your previous feedback you said to use token as the query string. I tried changing line 12 to const token = token and const token = parsed.token. But I got errors with both of those. So I changed it to parsed.t.  Not sure if I understood your feedback on this correctly.

File: frontend/src/components/NewPassword/NewPassword.js [hunk 11]
```diff
@@ -22,14 +27,34 @@ class NewPassword extends Component {
...
+        .then(this.handleRedirect)
+        .catch((error) => {
+          this.setState({ error: error.message });
+        });
+    }
+  };
+
+  handleRedirect = () => {
+    const { history } = this.props;
+    history.push('/login');
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: How about we just cut out this function and move the `history.push` up into the `handleSubmit` function. We can pull the `history` prop at the same time as the `token` prop on line 31.

File: frontend/src/components/NewPassword/NewPassword.js [hunk 12]
```diff
@@ -22,14 +27,34 @@ class NewPassword extends Component {
...
+    const { token } = this.props;
+    event.preventDefault();
+    const { password, confirmPassword } = this.state;
+    if (password !== confirmPassword) {
+      this.setState({ error: "Passwords don't match" });
+    } else {
+      resetPassword(token, password)
+        .then(this.handleRedirect)
+        .catch((error) => {
+          this.setState({ error: error.message });
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: I think the error is actually going to be on something like `error.response.data.errors[0].message`. `error.message` is probably just going to be "the request failed with XXX status." I think we may want to add some handling here, like if the error is a 400-level error, it indicates a bad token or bad password.

File: frontend/src/containers/ResetPassword/ResetPasswordPage.js [hunk 13]
```diff
@@ -1,23 +1,31 @@
...
-    <div className="passwordBackgroundImage">
-      <div className="newPasswordCardWrapper">
-        <NewPassword />
-        <div className="loginFooter row">
-          <Footer />
+const ResetPasswordPage = (props) => {
+  const { location } = props;
+  const parsed = querystring.parse(location.search.slice(1));
+  const { token } = parsed;
+  console.log(token);
```
Comment (vhoof) [inline frontend/src/containers/ResetPassword/ResetPasswordPage.js]: We probably don't want to log this. :)

File: frontend/src/containers/ResetRequest/ResetRequest.css [hunk 14]
```diff
@@ -0,0 +1,11 @@
+.btn-wrapper {
```
Comment (vhoof) [inline frontend/src/containers/ResetRequest/ResetRequest.css]: Can we make both of class names in here a little more specific to this page?

File: frontend/src/containers/ResetRequest/ResetRequest.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  handleChange = (event) => {
+    this.setState({ [event.target.name]: event.target.value });
+  };
+
+  handleSubmit = (event) => {
+    event.preventDefault();
+    const { email } = this.state;
+    requestPasswordReset(email)
+      .then(
+        this.setState({
```
Comment (vhoof) [inline frontend/src/containers/ResetRequest/ResetRequest.js]: I think we need an arrow function here, like `() => this.setState`...

File: frontend/src/lib/apiClient.js [hunk 16]
```diff
@@ -9,7 +9,10 @@ const apiClient = axios.create({
   baseURL,
 });
 
-// AUTHENTICATION
```
Comment (vhoof) [inline frontend/src/lib/apiClient.js]: Can we re-sort the functions in here into their categories and replace the category comments? It's hellishly hard to parse when the user and recipe and login functions are all jumbled together.

File: frontend/src/lib/apiClient.js [hunk 17]
```diff
@@ -27,48 +30,70 @@ const loginUser = (data) => {
...
 
-const adminEditUser = (data, userId) => apiClient.patch(`api/v1/users/${userId}`, {
-  first_name: data.firstName.value,
-  last_name: data.lastName.value,
-  email: data.email.value,
-}).then((response) => {
-  console.log(response);
-}).catch((error) => {
-  console.log(error.response.data);
+// Reset Password Function
```
Comment (vhoof) [inline frontend/src/lib/apiClient.js]: I think the function being called `resetPassword` is self-commenting. :)

File: frontend/src/lib/apiClient.js [hunk 18]
```diff
@@ -27,48 +30,70 @@ const loginUser = (data) => {
...
 
-// USERS
-
-const getUsers = () => apiClient.get('api/v1/users')
-  .then(results => results.data);
-
-const getUser = userId => apiClient.get(`api/v1/users/${userId}`, {
-  headers: {
-    Authorization: sessionStorage.jwt,
+// Reset Request
```
Comment (vhoof) [inline frontend/src/lib/apiClient.js]: Same here re: self-commenting.

**Comment by H12** [comment]
@andreaMT15  When you're done implementing @vhoof's feedback, please remove the "Changes Requested" label and `@` me in a comment so I know when it's ready for another look 👍

File: frontend/src/components/NewPassword/NewPassword.js [hunk 19]
```diff
@@ -22,14 +24,36 @@ class NewPassword extends Component {
...
+      resetPassword(token, password)
+        .then((response) => {
+          if (response) {
+            history.push('/login');
+          }
+        })
+        .catch((error) => {
+          if (error) {
+            this.setState({ error: 'Error' });
+            console.log('Error response', error.response.data.errors); // eslint no-console
```
Comment (andreaMT15) [inline frontend/src/components/NewPassword/NewPassword.js]: @vhoof @H12 I felt like leaving it more general to have a console log that gives us the error response was better for multiple types of errors rather than making it specific to the the reset password token error. This way if the error is unrelated to the password reset token, we will still get specific information for that particular error.

**Comment by andreaMT15** [comment]
@H12 this is ready for review again.

File: frontend/src/lib/apiClient.js [hunk 20]
```diff
@@ -9,7 +9,27 @@ const apiClient = axios.create({
   baseURL,
 });
 
-// AUTHENTICATION
```
Comment (vhoof) [inline frontend/src/lib/apiClient.js]: Is there any particular reason the comments were changed and all the functions were moved around? Seems like it'd be easier just to take what's in staging and add the two functions you need.

File: config/environments/development.rb [hunk 21]
```diff
@@ -4,6 +4,18 @@
...
+
+  config.action_mailer.smtp_settings = {
+    address: "smtp.gmail.com",
+    port: 587,
+    domain: "example.com",
+    authentication: "plain",
+    enable_starttls_auto: true,
+    user_name: ENV["GMAIL_USERNAME"],
+    password:ENV["GMAIL_PASSWORD"]
+  }
```
Comment (vhoof) [inline config/environments/development.rb]: Since we're not actually going to have gmail credentials on development machines, maybe we should remove all this configuration?

**Comment by andreaMT15** [comment]
api client file still needs to be fixed to match staging again. Will complete later this evening.

File: frontend/src/components/NewPassword/NewPassword.js [hunk 22]
```diff
@@ -22,17 +24,51 @@ class NewPassword extends Component {
...
+    const { isPasswordVisible, error } = this.state;
 
+    if (error) {
+      return (
+        <div className="errorMessage">
+          <p>
+            {error}
+            {' '}
+            <Link to="/resetrequest" id="requestLink">
+              Click here to request a new password reset token.
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: I might say "You can <Link>request a new password reset token</Link>." I think we can rely on most people knowing that you can click on links. :)

File: frontend/src/components/NewPassword/NewPassword.js [hunk 23]
```diff
@@ -22,17 +24,51 @@ class NewPassword extends Component {
...
+      this.setState({ error: "Passwords don't match" });
+    } else {
+      resetPassword(token, password)
+        .then(() => {
+          history.push('/login');
+        })
+        .catch((error) => {
+          if (error.response.data.errors.reset_password_token[0]) {
+            this.setState({
+              error: `Reset password token ${error.response.data.errors.reset_password_token[0]}.`,
```
Comment (vhoof) [inline frontend/src/components/NewPassword/NewPassword.js]: What does this actually end up returning?

File: app/controllers/api/v1/recipe_controller.rb [hunk 24]
```diff
@@ -0,0 +1,16 @@
+class Api::V1::RecipeController < ApplicationController
```
Comment (vhoof) [inline app/controllers/api/v1/recipe_controller.rb]: Where did this file come from?

File: frontend/package.json [hunk 25]
```diff
@@ -14,7 +14,7 @@
     "react-router-dom": "~4.3.1",
     "react-scripts": "~2.1.8",
     "reactstrap": "^7.1.0",
-    "universal-cookie": "~4.0.2"
+    "universal-cookie": "^4.0.2"
```
Comment (vhoof) [inline frontend/package.json]: I think we probably don't want to make changes to package.json/package-lock.json unless we actually changed some packages, so you can reset these changes.

File: frontend/src/components/NewPassword/NewPassword.js (hunk 23 above)
Comment (vhoof) [reply frontend/src/components/NewPassword/NewPassword.js]: I think we'll also want to handle the case where something goes wrong that isn't token-related. Like if the reset password token thing is there, then yeah, display this error. Else display something like, "Something went wrong with our system. Please try again."

File: frontend/src/components/NewPassword/NewPassword.js (hunk 22 above)
Comment (andreaMT15) [reply frontend/src/components/NewPassword/NewPassword.js]: @vhoof The reason I worded it that way is because when you look at it on the front end it's not obvious that it's a link. Or at least I didn't think it's obvious since it's the same color text :) I'll change it though. :)

File: app/controllers/api/v1/recipe_controller.rb (hunk 24 above)
Comment (andreaMT15) [reply app/controllers/api/v1/recipe_controller.rb]: @vhoof I'm not sure how this made it into our PR. Neither Fransisco or I have made changes to this file. All i can think of is that it got pushed because I merged staging into our branch.

File: frontend/package.json (hunk 25 above)
Comment (andreaMT15) [reply frontend/package.json]: @vhoof we kept getting errors that it couldn't find the universal-cookie package. So when we installed it, is when it made changes to package.json. How do we reset this without getting an error again?

Comment (vhoof) [reply frontend/package.json]: Yeah, the package was added in staging. Since it's already in package.json, all you need to do is `docker-compose build` to update your dependencies.

File: app/controllers/api/v1/recipe_controller.rb (hunk 24 above)
Comment (vhoof) [reply app/controllers/api/v1/recipe_controller.rb]: We should probably remove it. The actual file is `recipes_controller`, so not sure what this one is. And either way, it shouldn't be in your PR.

File: frontend/package.json (hunk 25 above)
Comment (vhoof) [reply frontend/package.json]: And you'll want to reset these files to their staging state first.

File: frontend/src/components/NewPassword/NewPassword.js (hunk 22 above)
Comment (vhoof) [reply frontend/src/components/NewPassword/NewPassword.js]: Ahh, I'd probably style it like the other links on the page (that blue or whatever).

**Comment by andreaMT15** [comment]
@vhoof This is finally ready for final review! :) I think we should finally be able to merge this now.
